### PR TITLE
[SW-2031] Expose Word2Vec to Python and Scala Algo API

### PR DESCRIPTION
**Pull request by jakubhava**

### CODE REVIEW

**Comment by jakubhava** [comment]
@mn-mikke depends on https://github.com/h2oai/h2o-3/pull/4882, before that change is merged, the CI will be red, but conceptually, this change is complete.

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala [hunk 1]
```diff
@@ -142,7 +145,8 @@ object Runner {
       ("H2OGLRM", classOf[GLRMParameters], "H2OUnsupervisedAlgorithm", Seq.empty),
       ("H2OGAM", classOf[GAMParameters], "H2OSupervisedAlgorithm", Seq.empty),
       ("H2OPCA", classOf[PCAParameters], "H2OUnsupervisedAlgorithm", Seq.empty),
-      ("H2OIsolationForest", classOf[IsolationForestParameters], "H2OTreeBasedUnsupervisedAlgorithm", Seq.empty))
+      ("H2OIsolationForest", classOf[IsolationForestParameters], "H2OTreeBasedUnsupervisedAlgorithm", Seq.empty),
+      ("H2OWord2Vec", classOf[Word2VecParameters], "H2OUnsupervisedAlgorithm", Seq("H2OWord2VecExtras")))
```
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: Is `H2OWord2Vec` an unsupervised H2O algorithm? Shouldn't it belong to the same category and namespace as `H2OTargetEncoder`?

File: core/src/main/scala/ai/h2o/sparkling/H2OFrame.scala [hunk 2]
```diff
@@ -51,6 +51,32 @@ class H2OFrame private (
...
+  def convertColumnsToStrings(columns: Array[String]): H2OFrame = {
+    val indices = columnNames.zipWithIndex.toMap
+    val selectedIndices = columns.map { name =>
+      indices.getOrElse(name, throw new IllegalArgumentException(s"Column $name does not exist in the frame $frameId"))
+    }
+    convertColumnsToStrings(selectedIndices)
+  }
+
+  def convertColumnsToStrings(columnIndices: Array[Int]): H2OFrame = {
+    val nonExisting = columnIndices.filterNot(columns.indices.contains(_))
```
Comment (mn-mikke) [inline core/src/main/scala/ai/h2o/sparkling/H2OFrame.scala:63]: nit: wouldn't work just `columnIndices.filterNot(columns.indices.contains)`

File: doc/src/site/sphinx/ml/sw_w2v.rst [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        Create the pipeline with the Word2Vec. You can configure all the available Word2Vec arguments using provided setters.
+
+        .. code:: python
+
+            from pysparkling.ml import H2OGBM, H2OWord2Vec
+            from pyspark.ml.feature import RegexTokenizer, StopWordsRemover
+
+            tokenizer = RegexTokenizer(inputCol="jobtitle", minTokenLength=2)
+            stopWordsRemover = StopWordsRemover(inputCol=tokenizer.getOutputCol())
+            w2v = H2OWord2Vec(sentSampleRate=0, epochs=0, featuresCol=stopWordsRemover.getOutputCol())
```
Comment (mn-mikke) [inline doc/src/site/sphinx/ml/sw_w2v.rst:107]: scala versions has `epochs` set to `10`

File: ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      .setInputCol("jobtitle")
+      .setMinTokenLength(2)
+
+    val stopWordsRemover = new StopWordsRemover()
+      .setInputCol(tokenizer.getOutputCol)
+
+    val w2v = new H2OWord2Vec()
+      .setSentSampleRate(0)
+      .setEpochs(10)
+      .setFeaturesCol(stopWordsRemover.getOutputCol)
```
Comment (mn-mikke) [inline ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala:51]: Shouldn't `H2OWord2Vec` Have `inputCol` and `outputCol` instead of `featuresCol` and `predictionCol`? as it's for `H2OTargetEncoder`?

File: doc/src/site/sphinx/ml/sw_w2v.rst (hunk 3 above)
Comment (jakubhava) [reply doc/src/site/sphinx/ml/sw_w2v.rst:107]: 👍

File: core/src/main/scala/ai/h2o/sparkling/H2OFrame.scala (hunk 2 above)
Comment (jakubhava) [reply core/src/main/scala/ai/h2o/sparkling/H2OFrame.scala:63]: Yes

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 1 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: H2OWord2Vec is unsupervised algorithm. But can move it

File: ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala (hunk 4 above)
Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala:51]: The code is based on H2O where Word 2 Vec is treated as algorithm. I can change though, don't have really strong preference

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 1 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: If I'm not mistaken Word2Vec is treated in H2O as an algorithm since there is no concept of transformers and estimators. IMHO, Word2Vec is a feature engineering stage that has to be combined with another algorithm. Spark categorizes as a feature estimator as well: https://spark.apache.org/docs/latest/ml-features.html#word2vec

cc @michalkurka

Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: ok

File: ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala (hunk 4 above)
Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala:51]: Removing featuresCol and predictionCol would mean that changes to the structure of H2OALgorithm classes etc as these parameters are defined there ( and Word2Vec depends on H2OAlgorithm)

I can change it, but would first merge PR in this state and iterate after in another PR

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 1 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: I can change it, but would first merge PR in this state and iterate after in another PR

Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: done, w2v in in the features namespace

File: ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala (hunk 4 above)
Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala:52]: I have exposed inputCol and outputCol and up-dated also the documentation. But still kept also params featuresCol and predictionCol available. Those can be removed in the follow-up PR, want to avoid doing that as part of this change.

Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OWord2VecTestSuite.scala:52]: because it shares all the pieces with H2OAlgorithm, mojo etc. Just prevents duplication the code.

I'm open to changing the structure as explained above, but I'm in favor of keeping it child-class of H2OAlgorithm ( since it is algorithm in H2O)

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 1 above)
Comment (michalkurka) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:149]: @mn-mikke @jakubhava great makes sense

**Comment by jakubhava** [comment]
@mn-mikke Can you please have a second go?

I have re-implemented this change so `H2OWord2Vec` does not depend on the `H2OAlgorithm` and is located in the features namespace

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    H2OModel(modelId)
+      .toMOJOModel(Identifiable.randomUID(classOf[H2OWord2Vec].getSimpleName), settings)
+  }
+
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  @DeveloperApi
+  override def transformSchema(schema: StructType): StructType = schema
```
Comment (mn-mikke) [inline ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:80]: Shouldn't be some parameter checks performed here? InputCol is set and in the dataset, column types, etc.

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    val params = getH2OAlgorithmParams(train) ++
+      Map("training_frame" -> train.frameId, "model_id" -> convertModelIdToKey(getModelId()))
+    val modelId = trainAndGetDestinationKey(s"/3/ModelBuilders/word2vec", params)
+    deleteRegisteredH2OFrames()
+    val settings = H2OMOJOSettings(predictionCol = getOutputCol)
+    H2OModel(modelId)
+      .toMOJOModel(Identifiable.randomUID(classOf[H2OWord2Vec].getSimpleName), settings)
+  }
+
+  def setInputCol(value: String): this.type = set(inputCol, value)
```
Comment (mn-mikke) [inline ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:75]: nit: Would it be possible to move setters closer to getters?

File: ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala [hunk 7]
```diff
@@ -51,4 +52,24 @@ trait EstimatorCommonUtils extends RestCommunication {
     target
   }
 
+  protected def convertModelIdToKey(key: String): String = {
+    if (H2OModel.modelExists(key)) {
+      val replacement = findAlternativeKey(key)
+      logWarning(
```
Comment (mn-mikke) [inline ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala:58]: In what scenario does this could happen?

Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala:58]: Couldn't the suffix be same UUID?

File: py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+from ai.h2o.sparkling import Initializer
+from ai.h2o.sparkling.ml.H2OStageBase import H2OStageBase
+from ai.h2o.sparkling.ml.Utils import Utils
+from ai.h2o.sparkling.ml.models import H2OMOJOModel
+from ai.h2o.sparkling.ml.params.H2OWord2VecParams import H2OWord2VecParams
+from pyspark import keyword_only
+from pyspark.ml.wrapper import JavaEstimator
+from pyspark.ml.param.shared import HasInputCol, HasOutputCol
+
+class H2OWord2Vec(HasInputCol, HasOutputCol, H2OWord2VecParams, H2OStageBase, JavaEstimator):
```
Comment (mn-mikke) [inline py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py:27]: Couldn't this class be generated?

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionWordEmbedding.scala [hunk 9]
```diff
@@ -30,13 +31,16 @@ trait H2OMOJOPredictionWordEmbedding {
     val schema = getWordEmbeddingPredictionSchema()
     val function = (r: Row) => {
       val model = H2OMOJOCache.getMojoBackend(uid, getMojo, this)
-      val pred = model.predictWord2Vec(RowConverter.toH2ORowData(r))
-      new GenericRowWithSchema(Array(pred.wordEmbeddings), schema)
+      val colIdx = model.m.getColIdx(getFeaturesCols().head)
```
Comment (mn-mikke) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionWordEmbedding.scala:34]: Is `inputCol` somewhere propagated to `featuresCols`?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/AlgorithmAPIRunner.scala [hunk 10]
```diff
@@ -92,4 +97,6 @@ object AlgorithmAPIRunner
       writeResultToFile(content, gridSearchAlgorithmContext, languageExtension, destinationDir)
     }
   }
+
+  private def generateWord2VecParams(): Unit = {}
```
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/AlgorithmAPIRunner.scala:101]: What is the purpose of method?

**Review by mn-mikke — COMMENTED** [review]
Would it possible to add some tess with a  specific values in asserts?

**Comment by mn-mikke** [comment]
@honza @michalkurka Can you review this PR as well?

File: py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py (hunk 8 above)
Comment (jakubhava) [reply py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py:27]: I had it generated in the previous attempt when it was based on H2OAlgorithm, but that attempt was discouraged. 

I do not like writing new generation logic just for this one instance. I would keep it like this

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/AlgorithmAPIRunner.scala (hunk 10 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/AlgorithmAPIRunner.scala:101]: None, will remove

File: ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala (hunk 7 above)
Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala:58]: I just moved this exact method without any changes from H2OAlgoritm, no changes in the method. 

Not sure if I understand the problem.

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionWordEmbedding.scala (hunk 9 above)
Comment (jakubhava) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionWordEmbedding.scala:34]: Good question, it is -> https://github.com/h2oai/sparkling-water/blob/3a86429e56e32d40d51e678ae3e7c0cfae0aecbe/scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala#L251

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  final val outputCol: Param[String] = new Param[String](this, "outputCol", "output column name")
+
+  setDefault(outputCol, "H2OWord2Vec_output")
+
+  /** @group getParam */
+  final def getOutputCol: String = $(outputCol)
+
+  def this() = this(Identifiable.randomUID(classOf[H2OWord2Vec].getSimpleName))
+
+  private def addValue: UserDefinedFunction = udf((array: Seq[String]) => array ++ Array(""))
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:46]: Good point, right now it fails. I will write tests for both empty array and null array with a proper exception for the user. Actually I propose internally just convert null to an empty array and have one exception.

Based on my tests, H2O does not support empty input for the H2O Word to Vec.

[the remark this answers is not in the document]
Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:46]: Implemented several tests to ensure we catch these edge cases 👍

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala (hunk 5 above)
Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:69]: Yup, completely missed this part. Thanks for checking. I will also add a test

File: py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py (hunk 8 above)
Comment (jakubhava) [reply py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py:27]: After giving it another thought, the generated code will be really simple, I will generate the python class to avoid manually typing the parameters.

**Comment by mn-mikke** [comment]
:-) my apologies @honza, I meant @honzasterba

File: py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py (hunk 8 above)
Comment (jakubhava) [reply py/src/ai/h2o/sparkling/ml/features/H2OWord2Vec.py:27]: Done

**Comment by jakubhava** [comment]
I hope that I have address most of the review suggestions, please have a look

File: ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala (hunk 7 above)
Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala:58]: ok, never mind. I just thought that purpose of this method is to avoid name collisions. I was wondering whether it couldn't be there an alternative approach.

Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/utils/EstimatorCommonUtils.scala:58]: Oki, thanks for the explanation. If you see a room for some improvement, please create a JIRA, and let's have a look at it as part of different PR. Thanks!

File: ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    val w2v = new H2OWord2Vec().setInputCol("value")
+
+    val thrown = intercept[IllegalArgumentException] {
+      w2v.fit(df)
+    }
+    assert(thrown.getMessage == "Empty DataFrame as an input for the H2OWord2Vec is not supported.")
+  }
+
+  test("Word2Vec with only one and empty input column") {
+
```
Comment (mn-mikke) [inline ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala:107]: nit: extra empty line

File: ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    val thrown = intercept[IllegalArgumentException] {
+      w2v.fit(df)
+    }
+    assert(thrown.getMessage == "Empty DataFrame as an input for the H2OWord2Vec is not supported.")
+  }
+
+  private def truncateAt(n: Double, p: Int): Double = { val s = math pow (10, p); (math floor n * s) / s }
+
+  test("Word2Vec with null input column") {
```
Comment (mn-mikke) [inline ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala:121]: Can you also add a similar test for a regular column?

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+  private def addValue: UserDefinedFunction =
+    udf((array: mutable.WrappedArray[String]) => {
+      array.toArray ++ Array("")
+    })
+
+  override def fit(dataset: Dataset[_]): H2OMOJOModel = {
+    val inputCol = getInputCol
+    val ds = dataset
+      .filter(inputCol + " is not null")
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:56]: To filter out `null` columns for the fitting phase. These are the cases you pass a null array as a row to the target encoder. H2O can't handle that so we filter it out

File: ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala (hunk 13 above)
Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala:121]: This test contains also that part if you look bellow:

It tests the output for the `null` row and also for the regular ``Array("how", "are", "you")`` row

Comment (mn-mikke) [reply ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala:121]: I see, thanks!

Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala:121]: Perfect!

File: ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala (hunk 12 above)
Comment (jakubhava) [reply ml/src/test/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTestSuite.scala:107]: Thanks!

File: booklet/src/sections/calling_h2o_algos.tex [hunk 15]
```diff
@@ -4,9 +4,16 @@
 
 	\section{Calling H2O Algorithms}
 
-	This section describes how to call H2O algorithms from Sparkling Water. In Scala and Python, Sparkling Water
+	This section describes how to call H2O algorithms and feature transformers from Sparkling Water. In Scala and Python, Sparkling Water
 	exposes H2O algorithms via its own API. In R, we still need to use H2O's R API.
 
+	Following features are exposed in Sparkling Water:
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply booklet/src/sections/calling_h2o_algos.tex:10]: Thanks!

File: api-generation/build.gradle [hunk 16]
```diff
@@ -8,6 +8,9 @@ dependencies {
   implementation("ai.h2o:h2o-algos:${h2oVersion}")
   implementation("ai.h2o:h2o-automl:${h2oVersion}")
   implementation("ai.h2o:h2o-ext-xgboost:${h2oVersion}")
+  compileOnly("org.apache.spark:spark-core_${scalaBaseVersion}:${sparkVersion}")
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply api-generation/build.gradle:11]: Actually only one is required, thanks for checking that. I need `org.apache.spark:spark-mllib` because I use the `Params` class at https://github.com/h2oai/sparkling-water/blob/5cafdc31603d6cb4c4e15687fdc3066566864e62/api-generation/src/main/scala/ai/h2o/sparkling/api/generation/python/Word2VecTemplate.scala#L20

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:56]: Oh ok, I got confused by string concatenation. Is there any reason for not using functions from `import org.apache.spark.sql.functions`?

File: api-generation/build.gradle (hunk 16 above)
[the remark this answers is not in the document]
Comment (mn-mikke) [reply api-generation/build.gradle]: ok, thanks!

File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (jakubhava) [reply ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2Vec.scala:56]: Not at all, found out we can do `filter(col(inputCol).isNotNull)` which is nicer anyways. Thanks!

**Review by mn-mikke — APPROVED** [review]
LGTM, thanks! @michalkurka @honzasterba Can you take a look at the PR at least from a high level point of view?

**Review by honzasterba — APPROVED** [review]
lgtm

**Comment by jakubhava** [comment]
Thank you @mn-mikke for the thorough review! We are green and finally merging 👍
